### PR TITLE
Improvements to signature help (especially when fix_pos=true)

### PR DESCRIPTION
**Pull request by HansPinckaers**
Hi,

These are quite some significant changes to make the signature help pop-up more reliably. Also improves the positioning of the signature help.

Improvements:
- When entering insert mode always show signature help when available
- When possible, request the signature at the nearest trigger character, this improves reliability. For example, when entering insert mode inside a function call, but not next to a trigger character, the signature will now be shown.
- Make the LSP responsible to close the signature help (thus when none are returned and fix_pos==true)
- When the LSP returns `<function_name>(...)` also use `function_name` to position the signature help more correctly. This is useful if there are more signatures on a single line.

Most of these improvements have only been tested during my main development work. Where I have `fix_pos==true`. Testing is needed with `fix_pos==false`.

### CODE REVIEW

**Comment by dsully** [comment]
Would love to see these changes merged.

File: lua/lsp_signature.lua [hunk 1]
```diff
@@ -212,12 +211,24 @@ local function signature_handler(err, method, result, client_id, bufnr, config)
...
+        local function_on_line = line_to_cursor:match('.*' .. function_name)
+        if function_on_line then
+          woff = #line_to_cursor - #function_on_line + #function_name
+        else
+          woff = sig_woff + (#line_to_cursor - woff)
+        end
         woff = -woff
       else
-        woff = -3
+        woff = 3
```
Comment (ray-x) [inline lua/lsp_signature.lua:231]: Should be -3 by default.

**Review by ray-x — APPROVED** [review]
Thanks for the PR.  It looks great.

**Comment by HansPinckaers** [comment]
Thanks for merging this! I wanted to do a bit more testing, but it already works pretty ok. I do have an issue where sometimes the signature stays visible going to normal mode. Pretty sure this is some kind of race condition.

**Comment by ray-x** [comment]
I feel it might be better to merge it. As it has lots of code turmoils and it will be painful for you to maintain this branch.

**Comment by HansPinckaers** [comment]
Good point, thanks for merging!

**Pull request merged**